File: packages/engine/src/cpus/6502/instructions/cpy.js
import { AddressingMode, CycleType } from '..'

/**
 * ComPare Y register
 *
 * Z,C,N = Y-M
 *
 * Affects Flags: N Z C
 *
 * @example
 * MODE          SYNTAX        HEX LEN TIM
 * Immediate     CPY #$44      $C0  2   2
 * Zero Page     CPY $44       $C4  2   3
 * Absolute      CPY $4400     $CC  3   4
 */
export function* cpy(mode, addr) {
  const a = this.y
  const b = mode === AddressingMode.IMM ? addr : this.read(Array.isArray(addr) ? addr[0] : addr)
  this.carry = a >= b
  this.zero = a === b
  let v = a - b
  if (v < 0) v = 256 + v
  this.negative = (v & 0x80) > 0
  yield CycleType.GET
}

File: packages/engine/src/cores/famibob/cpu/instructions/cpy.zig
const std = @import("std");
const CPU = @import("../cpu.zig").CPU;
const Addressing = @import("../opcode.zig").Addressing;

// ComPare Y register
//
// Z,C,N = Y-M
//
// Affects Flags: N Z C
//
// MODE          SYNTAX        HEX LEN TIM
// Immediate     CPY #$44      $C0  2   2
// Zero Page     CPY $44       $C4  2   3
// Absolute      CPY $4400     $CC  3   4

pub fn cpy(self: *CPU, addressing: Addressing) void {
    const a = self.y;
    var b: u8 = 0;

    switch (addressing) {
        .imm => |v| {
            b = v;
        },
        .zpg => |v| {
            b = self.read(v);
        },
        .abs => |v| {
            b = self.read(v);
        },
        else => {},
    }

    self.c = a >= b;
    self.z = a == b;
    const v = a -% b;
    self.n = (v & 0x80) > 0;
}
